File: nasdaq-cse/rms/manager.py
"""
Risk Management System (RMS) for monitoring positions and margins
"""
from datetime import datetime
from typing import Dict, List, Optional, Tuple
from sqlalchemy.orm import Session
from core.models import User, Position, Contract
from storage.database import db_manager
import asyncio


class RiskManager:
    """
    Comprehensive risk management system for monitoring and controlling trading risks
    """
    
    def __init__(self):
        self.risk_limits = {
            'max_position_size': 1000,  # Max position size per contract
            'max_total_exposure': 500000,  # Max total exposure per user
            'margin_call_threshold': 0.8,  # Margin call at 80% utilization
            'force_liquidation_threshold': 0.95,  # Force liquidation at 95%
            'max_daily_loss': 0.05,  # Max 5% daily loss
            'concentration_limit': 0.3,  # Max 30% in single position
            'volatility_limit': 0.1  # Max 10% daily volatility exposure
        }
        self.risk_alerts = []
    
    async def check_pre_trade_risk(self, user_id: int, order_request: Dict) -> Dict:
        """
        Check risk limits before allowing a trade
        """
        db = next(db_manager.get_db())
        
        try:
            user = db.query(User).filter(User.id == user_id).first()
            if not user:
                return {'allowed': False, 'reason': 'User not found'}
            
            # Get user's current positions
            positions = db.query(Position).filter(Position.user_id == user_id).all()
            
            # Calculate current exposure
            total_exposure = sum(abs(pos.quantity * pos.avg_entry_price) for pos in positions)
            
            # Check position size limit
            new_exposure = float(order_request['quantity']) * float(order_request.get('price', 2000))
            if new_exposure > self.risk_limits['max_position_size'] * 2000:  # Assuming $2000 per unit
                return {'allowed': False, 'reason': 'Position size exceeds limit'}
            
            # Check total exposure limit
            if total_exposure + new_exposure > self.risk_limits['max_total_exposure']:
                return {'allowed': False, 'reason': 'Total exposure would exceed limit'}
            
            # Check margin requirements
            margin_check = await self._check_margin_requirements(user, positions, order_request, db)
            if not margin_check['sufficient']:
                return {'allowed': False, 'reason': margin_check['reason']}
            
            # Check concentration limits
            concentration_check = await self._check_concentration_limits(positions, order_request)
            if not concentration_check['allowed']:
                return {'allowed': False, 'reason': concentration_check['reason']}
            
            return {'allowed': True, 'reason': 'Risk checks passed'}
            
        finally:
            db.close()
    
    async def check_post_trade_risk(self, user_id: int) -> Dict:
        """
        Check risk metrics after a trade is executed
        """
        db = next(db_manager.get_db())
        
        try:
            user = db.query(User).filter(User.id == user_id).first()
            positions = db.query(Position).filter(Position.user_id == user_id).all()
            
            risk_metrics = await self._calculate_risk_metrics(user, positions)
            alerts = await self._generate_risk_alerts(risk_metrics)
            
            return {
                'user_id': user_id,
                'risk_metrics': risk_metrics,
                'alerts': alerts,
                'timestamp': datetime.utcnow().isoformat()
            }
            
        finally:
            db.close()
    
    async def monitor_margin_requirements(self, user_id: int, current_prices: Dict[int, float]) -> Dict:
        """
        Monitor margin requirements and generate margin calls if needed
        """
        db = next(db_manager.get_db())
        
        try:
            user = db.query(User).filter(User.id == user_id).first()
            positions = db.query(Position).filter(Position.user_id == user_id).all()
            
            if not user or not positions:
                return {'margin_adequate': True, 'margin_call': False}
            
            # Calculate total margin requirement
            total_margin_required = 0
            total_unrealized_pnl = 0
            
            for position in positions:
                if position.contract_id in current_prices:
                    current_price = current_prices[position.contract_id]
                    
                    # Update unrealized P&L
                    if position.quantity != 0:
                        position.unrealized_pnl = (current_price - position.avg_entry_price) * position.quantity
                        total_unrealized_pnl += position.unrealized_pnl
                    
                    # Calculate margin requirement
                    contract = db.query(Contract).filter(Contract.id == position.contract_id).first()
                    if contract:
                        margin_per_unit = contract.maintenance_margin
                        total_margin_required += abs(position.quantity) * margin_per_unit
            
            # Calculate available margin
            account_equity = user.account_balance + total_unrealized_pnl
            available_margin = account_equity - total_margin_required
            margin_utilization = total_margin_required / account_equity if account_equity > 0 else 1.0
            
            margin_call = margin_utilization > self.risk_limits['margin_call_threshold']
            force_liquidation = margin_utilization > self.risk_limits['force_liquidation_threshold']
            
            # Update user's margin available
            user.margin_available = max(0, available_margin)
            db.commit()
            
            return {
                'margin_adequate': not margin_call,
                'margin_call': margin_call,
                'force_liquidation': force_liquidation,
                'margin_utilization': margin_utilization,
                'total_margin_required': total_margin_required,
                'available_margin': available_margin,
                'account_equity': account_equity
            }
            
        finally:
            db.close()
    
    async def calculate_var(self, user_id: int, confidence_level: float = 0.95, time_horizon: int = 1) -> Dict:
        """
        Calculate Value at Risk (VaR) for user's portfolio
        """
        db = next(db_manager.get_db())
        
        try:
            positions = db.query(Position).filter(Position.user_id == user_id).all()
            
            if not positions:
                return {'var': 0.0, 'expected_shortfall': 0.0}
            
            # Simplified VaR calculation
            total_exposure = sum(abs(pos.quantity * pos.avg_entry_price) for pos in positions)
            
            # Assume 2% daily volatility for gold
            daily_volatility = 0.02
            
            # VaR calculation (parametric method)
            from scipy.stats import norm
            z_score = norm.ppf(1 - confidence_level)
            var = total_exposure * daily_volatility * z_score * (time_horizon ** 0.5)
            
            # Expected Shortfall (Conditional VaR)
            expected_shortfall = total_exposure * daily_volatility * norm.pdf(z_score) / (1 - confidence_level)
            
            return {
                'var': abs(var),
                'expected_shortfall': abs(expected_shortfall),
                'confidence_level': confidence_level,
                'time_horizon_days': time_horizon,
                'total_exposure': total_exposure
            }
            
        except ImportError:
            # Fallback if scipy is not available
            total_exposure = sum(abs(pos.quantity * pos.avg_entry_price) for pos in positions)
            var = total_exposure * 0.02 * 1.65  # Approximate 95% VaR
            return {
                'var': var,
                'expected_shortfall': var * 1.3,
                'confidence_level': confidence_level,
                'time_horizon_days': time_horizon,
                'total_exposure': total_exposure
            }
        finally:
            db.close()
    
    async def generate_risk_report(self, user_id: int, current_prices: Dict[int, float]) -> Dict:
        """
        Generate comprehensive risk report for a user
        """
        margin_status = await self.monitor_margin_requirements(user_id, current_prices)
        var_metrics = await self.calculate_var(user_id)
        post_trade_risk = await self.check_post_trade_risk(user_id)
        
        risk_score = self._calculate_risk_score(margin_status, var_metrics, post_trade_risk)
        
        return {
            'user_id': user_id,
            'timestamp': datetime.utcnow().isoformat(),
            'risk_score': risk_score,
            'margin_status': margin_status,
            'var_metrics': var_metrics,
            'risk_metrics': post_trade_risk.get('risk_metrics', {}),
            'alerts': post_trade_risk.get('alerts', []),
            'recommendations': self._generate_risk_recommendations(risk_score, margin_status, var_metrics)
        }
    
    async def _check_margin_requirements(self, user: object, positions: List, order_request: Dict, db: Session) -> Dict:
        """
        Check if user has sufficient margin for new order
        """
        # Get contract for margin calculation
        contract = db.query(Contract).filter(Contract.symbol == order_request['contract_symbol']).first()
        if not contract:
            return {'sufficient': False, 'reason': 'Contract not found'}
        
        # Calculate additional margin needed
        additional_margin = float(order_request['quantity']) * contract.initial_margin
        
        if user.margin_available < additional_margin:
            return {'sufficient': False, 'reason': 'Insufficient margin available'}
        
        return {'sufficient': True, 'reason': 'Margin requirements met'}
    
    async def _check_concentration_limits(self, positions: List, order_request: Dict) -> Dict:
        """
        Check position concentration limits
        """
        total_exposure = sum(abs(pos.quantity * pos.avg_entry_price) for pos in positions)
        
        if total_exposure == 0:
            return {'allowed': True, 'reason': 'No existing positions'}
        
        # Calculate exposure for the contract in the order
        contract_symbol = order_request['contract_symbol']
        contract_exposure = sum(
            abs(pos.quantity * pos.avg_entry_price) for pos in positions
            # Note: This is simplified - in reality we'd match by contract_id
        )
        
        new_exposure = float(order_request['quantity']) * float(order_request.get('price', 2000))
        projected_contract_exposure = contract_exposure + new_exposure
        projected_concentration = projected_contract_exposure / (total_exposure + new_exposure)
        
        if projected_concentration > self.risk_limits['concentration_limit']:
            return {
                'allowed': False,
                'reason': f'Position concentration would exceed {self.risk_limits["concentration_limit"]*100}% limit'
            }
        
        return {'allowed': True, 'reason': 'Concentration limits satisfied'}
    
    async def _calculate_risk_metrics(self, user: object, positions: List) -> Dict:
        """
        Calculate comprehensive risk metrics
        """
        if not positions:
            return {
                'total_exposure': 0,
                'leverage_ratio': 0,
                'concentration_ratio': 0,
                'unrealized_pnl': 0,
                'position_count': 0
            }
        
        total_exposure = sum(abs(pos.quantity * pos.avg_entry_price) for pos in positions)
        total_unrealized_pnl = sum(pos.unrealized_pnl for pos in positions)
        leverage_ratio = total_exposure / user.account_balance if user.account_balance > 0 else 0
        
        # Calculate concentration (largest position as % of total)
        position_exposures = [abs(pos.quantity * pos.avg_entry_price) for pos in positions]
        max_position_exposure = max(position_exposures) if position_exposures else 0
        concentration_ratio = max_position_exposure / total_exposure if total_exposure > 0 else 0
        
        return {
            'total_exposure': total_exposure,
            'leverage_ratio': leverage_ratio,
            'concentration_ratio': concentration_ratio,
            'unrealized_pnl': total_unrealized_pnl,
            'position_count': len(positions),
            'largest_position': max_position_exposure
        }
    
    async def _generate_risk_alerts(self, risk_metrics: Dict) -> List[Dict]:
        """
        Generate risk alerts based on metrics
        """
        alerts = []
        
        if risk_metrics['leverage_ratio'] > 5:
            alerts.append({
                'type': 'HIGH_LEVERAGE',
                'severity': 'HIGH',
                'message': f"Leverage ratio {risk_metrics['leverage_ratio']:.1f}x is very high",
                'recommendation': 'Consider reducing position sizes'
            })
        
        if risk_metrics['concentration_ratio'] > self.risk_limits['concentration_limit']:
            alerts.append({
                'type': 'HIGH_CONCENTRATION',
                'severity': 'MEDIUM',
                'message': f"Position concentration {risk_metrics['concentration_ratio']:.1%} exceeds limit",
                'recommendation': 'Diversify positions across multiple contracts'
            })
        
        if risk_metrics['unrealized_pnl'] < -risk_metrics.get('total_exposure', 0) * 0.1:
            alerts.append({
                'type': 'LARGE_UNREALIZED_LOSS',
                'severity': 'HIGH',
                'message': 'Unrealized losses exceed 10% of exposure',
                'recommendation': 'Review positions and consider stop-loss orders'
            })
        
        return alerts
    
    def _calculate_risk_score(self, margin_status: Dict, var_metrics: Dict, post_trade_risk: Dict) -> float:
        """
        Calculate overall risk score (0-100, higher is riskier)
        """
        score = 0
        
        # Margin utilization component (0-40 points)
        margin_util = margin_status.get('margin_utilization', 0)
        score += min(40, margin_util * 50)
        
        # VaR component (0-30 points)
        var_ratio = var_metrics.get('var', 0) / max(var_metrics.get('total_exposure', 1), 1)
        score += min(30, var_ratio * 1000)
        
        # Risk metrics component (0-30 points)
        risk_metrics = post_trade_risk.get('risk_metrics', {})
        leverage = risk_metrics.get('leverage_ratio', 0)
        concentration = risk_metrics.get('concentration_ratio', 0)
        score += min(15, leverage * 3) + min(15, concentration * 50)
        
        return min(100, score)
    
    def _generate_risk_recommendations(self, risk_score: float, margin_status: Dict, var_metrics: Dict) -> List[str]:
        """
        Generate risk management recommendations
        """
        recommendations = []
        
        if risk_score > 70:
            recommendations.append("HIGH RISK: Immediate action required to reduce risk exposure")
        elif risk_score > 50:
            recommendations.append("MEDIUM RISK: Monitor positions closely and consider risk reduction")
        
        if margin_status.get('margin_utilization', 0) > 0.8:
            recommendations.append("Consider adding funds or reducing positions to improve margin situation")
        
        if var_metrics.get('var', 0) > 10000:
            recommendations.append("High VaR detected - consider portfolio hedging strategies")
        
        return recommendations


# Global instance
risk_manager = RiskManager()

File: nasdaq-cse-go/internal/rms/manager.go
// Package rms provides Risk Management System functionality
package rms

import (
	"fmt"
	"math"
	"time"

	"github.com/smaruf/python-ai-course/nasdaq-cse-go/internal/core"
	"gorm.io/gorm"
)

// RiskLimits defines various risk limits for trading
type RiskLimits struct {
	MaxPositionSize           float64 // Max position size per contract
	MaxTotalExposure          float64 // Max total exposure per user
	MarginCallThreshold       float64 // Margin call at 80% utilization
	ForceLiquidationThreshold float64 // Force liquidation at 95%
	MaxDailyLoss              float64 // Max 5% daily loss
	ConcentrationLimit        float64 // Max 30% in single position
	VolatilityLimit           float64 // Max 10% daily volatility exposure
}

// DefaultRiskLimits returns default risk limits
func DefaultRiskLimits() RiskLimits {
	return RiskLimits{
		MaxPositionSize:           1000,
		MaxTotalExposure:          500000,
		MarginCallThreshold:       0.8,
		ForceLiquidationThreshold: 0.95,
		MaxDailyLoss:              0.05,
		ConcentrationLimit:        0.3,
		VolatilityLimit:           0.1,
	}
}

// RiskAlert represents a risk alert
type RiskAlert struct {
	Type           string `json:"type"`
	Severity       string `json:"severity"`
	Message        string `json:"message"`
	Recommendation string `json:"recommendation"`
}

// RiskMetrics contains calculated risk metrics
type RiskMetrics struct {
	TotalExposure      float64 `json:"total_exposure"`
	LeverageRatio      float64 `json:"leverage_ratio"`
	ConcentrationRatio float64 `json:"concentration_ratio"`
	UnrealizedPnL      float64 `json:"unrealized_pnl"`
	PositionCount      int     `json:"position_count"`
	LargestPosition    float64 `json:"largest_position"`
}

// VaRMetrics contains Value at Risk calculations
type VaRMetrics struct {
	VaR               float64 `json:"var"`
	ExpectedShortfall float64 `json:"expected_shortfall"`
	ConfidenceLevel   float64 `json:"confidence_level"`
	TimeHorizonDays   int     `json:"time_horizon_days"`
	TotalExposure     float64 `json:"total_exposure"`
}

// MarginStatus contains margin monitoring information
type MarginStatus struct {
	MarginAdequate       bool    `json:"margin_adequate"`
	MarginCall           bool    `json:"margin_call"`
	ForceLiquidation     bool    `json:"force_liquidation"`
	MarginUtilization    float64 `json:"margin_utilization"`
	TotalMarginRequired  float64 `json:"total_margin_required"`
	AvailableMargin      float64 `json:"available_margin"`
	AccountEquity        float64 `json:"account_equity"`
}

// RiskManager manages trading risks and monitoring
type RiskManager struct {
	db         *gorm.DB
	riskLimits RiskLimits
	riskAlerts []RiskAlert
}

// NewRiskManager creates a new risk manager
func NewRiskManager(db *gorm.DB) *RiskManager {
	return &RiskManager{
		db:         db,
		riskLimits: DefaultRiskLimits(),
		riskAlerts: make([]RiskAlert, 0),
	}
}

// CheckPreTradeRisk checks risk limits before allowing a trade
func (rm *RiskManager) CheckPreTradeRisk(userID uint, orderRequest core.OrderCreateRequest) map[string]interface{} {
	var user core.User
	if err := rm.db.First(&user, userID).Error; err != nil {
		return map[string]interface{}{
			"allowed": false,
			"reason":  "User not found",
		}
	}

	// Get user's current positions
	var positions []core.Position
	rm.db.Where("user_id = ?", userID).Find(&positions)

	// Calculate current exposure
	totalExposure := 0.0
	for _, pos := range positions {
		totalExposure += math.Abs(pos.Quantity * pos.AvgEntryPrice)
	}

	// Check position size limit
	estimatedPrice := 2000.0 // Default price estimate
	newExposure := orderRequest.Quantity * estimatedPrice
	if newExposure > rm.riskLimits.MaxPositionSize*estimatedPrice {
		return map[string]interface{}{
			"allowed": false,
			"reason":  "Position size exceeds limit",
		}
	}

	// Check total exposure limit
	if totalExposure+newExposure > rm.riskLimits.MaxTotalExposure {
		return map[string]interface{}{
			"allowed": false,
			"reason":  "Total exposure would exceed limit",
		}
	}

	// Check margin requirements
	marginCheck := rm.checkMarginRequirements(&user, positions, orderRequest)
	if !marginCheck["sufficient"].(bool) {
		return map[string]interface{}{
			"allowed": false,
			"reason":  marginCheck["reason"],
		}
	}

	// Check concentration limits
	concentrationCheck := rm.checkConcentrationLimits(positions, orderRequest)
	if !concentrationCheck["allowed"].(bool) {
		return map[string]interface{}{
			"allowed": false,
			"reason":  concentrationCheck["reason"],
		}
	}

	return map[string]interface{}{
		"allowed": true,
		"reason":  "Risk checks passed",
	}
}

// CheckPostTradeRisk checks risk metrics after a trade is executed
func (rm *RiskManager) CheckPostTradeRisk(userID uint) map[string]interface{} {
	var user core.User
	if err := rm.db.First(&user, userID).Error; err != nil {
		return map[string]interface{}{
			"error": "User not found",
		}
	}

	var positions []core.Position
	rm.db.Where("user_id = ?", userID).Find(&positions)

	riskMetrics := rm.calculateRiskMetrics(&user, positions)
	alerts := rm.generateRiskAlerts(riskMetrics)

	return map[string]interface{}{
		"user_id":      userID,
		"risk_metrics": riskMetrics,
		"alerts":       alerts,
		"timestamp":    time.Now().Format(time.RFC3339),
	}
}

// MonitorMarginRequirements monitors margin requirements and generates margin calls
func (rm *RiskManager) MonitorMarginRequirements(userID uint, currentPrices map[uint]float64) MarginStatus {
	var user core.User
	if err := rm.db.First(&user, userID).Error; err != nil {
		return MarginStatus{MarginAdequate: true, MarginCall: false}
	}

	var positions []core.Position
	rm.db.Where("user_id = ?", userID).Find(&positions)

	if len(positions) == 0 {
		return MarginStatus{MarginAdequate: true, MarginCall: false}
	}

	// Calculate total margin requirement
	totalMarginRequired := 0.0
	totalUnrealizedPnL := 0.0

	for _, position := range positions {
		if currentPrice, exists := currentPrices[position.ContractID]; exists {
			// Update unrealized P&L
			if position.Quantity != 0 {
				position.UnrealizedPnL = (currentPrice - position.AvgEntryPrice) * position.Quantity
				totalUnrealizedPnL += position.UnrealizedPnL
			}

			// Get contract for margin calculation
			var contract core.Contract
			if err := rm.db.First(&contract, position.ContractID).Error; err == nil {
				marginPerUnit := contract.MaintenanceMargin
				totalMarginRequired += math.Abs(position.Quantity) * marginPerUnit
			}
		}
	}

	// Calculate available margin
	accountEquity := user.AccountBalance + totalUnrealizedPnL
	availableMargin := accountEquity - totalMarginRequired
	var marginUtilization float64
	if accountEquity > 0 {
		marginUtilization = totalMarginRequired / accountEquity
	} else {
		marginUtilization = 1.0
	}

	marginCall := marginUtilization > rm.riskLimits.MarginCallThreshold
	forceLiquidation := marginUtilization > rm.riskLimits.ForceLiquidationThreshold

	// Update user's margin available
	user.MarginAvailable = math.Max(0, availableMargin)
	rm.db.Save(&user)

	return MarginStatus{
		MarginAdequate:       !marginCall,
		MarginCall:           marginCall,
		ForceLiquidation:     forceLiquidation,
		MarginUtilization:    marginUtilization,
		TotalMarginRequired:  totalMarginRequired,
		AvailableMargin:      availableMargin,
		AccountEquity:        accountEquity,
	}
}

// CalculateVaR calculates Value at Risk for user's portfolio
func (rm *RiskManager) CalculateVaR(userID uint, confidenceLevel float64, timeHorizon int) VaRMetrics {
	var positions []core.Position
	rm.db.Where("user_id = ?", userID).Find(&positions)

	if len(positions) == 0 {
		return VaRMetrics{
			VaR:               0.0,
			ExpectedShortfall: 0.0,
			ConfidenceLevel:   confidenceLevel,
			TimeHorizonDays:   timeHorizon,
		}
	}

	// Calculate total exposure
	totalExposure := 0.0
	for _, pos := range positions {
		totalExposure += math.Abs(pos.Quantity * pos.AvgEntryPrice)
	}

	// Simplified VaR calculation (parametric method)
	// Assume 2% daily volatility for gold
	dailyVolatility := 0.02

	// For 95% confidence level, z-score is approximately 1.65
	// For 99% confidence level, z-score is approximately 2.33
	var zScore float64
	switch {
	case confidenceLevel >= 0.99:
		zScore = 2.33
	case confidenceLevel >= 0.95:
		zScore = 1.65
	default:
		zScore = 1.28 // 90% confidence
	}

	// VaR calculation
	var1Day := totalExposure * dailyVolatility * zScore
	varTimeHorizon := var1Day * math.Sqrt(float64(timeHorizon))

	// Expected Shortfall (simplified)
	expectedShortfall := varTimeHorizon * 1.3

	return VaRMetrics{
		VaR:               math.Abs(varTimeHorizon),
		ExpectedShortfall: math.Abs(expectedShortfall),
		ConfidenceLevel:   confidenceLevel,
		TimeHorizonDays:   timeHorizon,
		TotalExposure:     totalExposure,
	}
}

// GenerateRiskReport generates comprehensive risk report for a user
func (rm *RiskManager) GenerateRiskReport(userID uint, currentPrices map[uint]float64) map[string]interface{} {
	marginStatus := rm.MonitorMarginRequirements(userID, currentPrices)
	varMetrics := rm.CalculateVaR(userID, 0.95, 1)
	postTradeRisk := rm.CheckPostTradeRisk(userID)

	riskScore := rm.calculateRiskScore(marginStatus, varMetrics, postTradeRisk)

	return map[string]interface{}{
		"user_id":         userID,
		"timestamp":       time.Now().Format(time.RFC3339),
		"risk_score":      riskScore,
		"margin_status":   marginStatus,
		"var_metrics":     varMetrics,
		"risk_metrics":    postTradeRisk["risk_metrics"],
		"alerts":          postTradeRisk["alerts"],
		"recommendations": rm.generateRiskRecommendations(riskScore, marginStatus, varMetrics),
	}
}

// checkMarginRequirements checks if user has sufficient margin for new order
func (rm *RiskManager) checkMarginRequirements(user *core.User, positions []core.Position, orderRequest core.OrderCreateRequest) map[string]interface{} {
	// Get contract for margin calculation
	var contract core.Contract
	if err := rm.db.Where("symbol = ?", orderRequest.ContractSymbol).First(&contract).Error; err != nil {
		return map[string]interface{}{
			"sufficient": false,
			"reason":     "Contract not found",
		}
	}

	// Calculate additional margin needed
	additionalMargin := orderRequest.Quantity * contract.InitialMargin

	if user.MarginAvailable < additionalMargin {
		return map[string]interface{}{
			"sufficient": false,
			"reason":     "Insufficient margin available",
		}
	}

	return map[string]interface{}{
		"sufficient": true,
		"reason":     "Margin requirements met",
	}
}

// checkConcentrationLimits checks position concentration limits
func (rm *RiskManager) checkConcentrationLimits(positions []core.Position, orderRequest core.OrderCreateRequest) map[string]interface{} {
	totalExposure := 0.0
	for _, pos := range positions {
		totalExposure += math.Abs(pos.Quantity * pos.AvgEntryPrice)
	}

	if totalExposure == 0 {
		return map[string]interface{}{
			"allowed": true,
			"reason":  "No existing positions",
		}
	}

	// Calculate exposure for the contract in the order (simplified)
	contractExposure := 0.0
	estimatedPrice := 2000.0 // Default price estimate
	
	newExposure := orderRequest.Quantity * estimatedPrice
	projectedContractExposure := contractExposure + newExposure
	projectedConcentration := projectedContractExposure / (totalExposure + newExposure)

	if projectedConcentration > rm.riskLimits.ConcentrationLimit {
		return map[string]interface{}{
			"allowed": false,
			"reason":  fmt.Sprintf("Position concentration would exceed %.0f%% limit", rm.riskLimits.ConcentrationLimit*100),
		}
	}

	return map[string]interface{}{
		"allowed": true,
		"reason":  "Concentration limits satisfied",
	}
}

// calculateRiskMetrics calculates comprehensive risk metrics
func (rm *RiskManager) calculateRiskMetrics(user *core.User, positions []core.Position) RiskMetrics {
	if len(positions) == 0 {
		return RiskMetrics{
			TotalExposure:      0,
			LeverageRatio:      0,
			ConcentrationRatio: 0,
			UnrealizedPnL:      0,
			PositionCount:      0,
		}
	}

	totalExposure := 0.0
	totalUnrealizedPnL := 0.0
	var positionExposures []float64

	for _, pos := range positions {
		exposure := math.Abs(pos.Quantity * pos.AvgEntryPrice)
		totalExposure += exposure
		totalUnrealizedPnL += pos.UnrealizedPnL
		positionExposures = append(positionExposures, exposure)
	}

	var leverageRatio float64
	if user.AccountBalance > 0 {
		leverageRatio = totalExposure / user.AccountBalance
	}

	// Calculate concentration (largest position as % of total)
	maxPositionExposure := 0.0
	for _, exposure := range positionExposures {
		if exposure > maxPositionExposure {
			maxPositionExposure = exposure
		}
	}

	var concentrationRatio float64
	if totalExposure > 0 {
		concentrationRatio = maxPositionExposure / totalExposure
	}

	return RiskMetrics{
		TotalExposure:      totalExposure,
		LeverageRatio:      leverageRatio,
		ConcentrationRatio: concentrationRatio,
		UnrealizedPnL:      totalUnrealizedPnL,
		PositionCount:      len(positions),
		LargestPosition:    maxPositionExposure,
	}
}

// generateRiskAlerts generates risk alerts based on metrics
func (rm *RiskManager) generateRiskAlerts(riskMetrics RiskMetrics) []RiskAlert {
	var alerts []RiskAlert

	if riskMetrics.LeverageRatio > 5 {
		alerts = append(alerts, RiskAlert{
			Type:           "HIGH_LEVERAGE",
			Severity:       "HIGH",
			Message:        fmt.Sprintf("Leverage ratio %.1fx is very high", riskMetrics.LeverageRatio),
			Recommendation: "Consider reducing position sizes",
		})
	}

	if riskMetrics.ConcentrationRatio > rm.riskLimits.ConcentrationLimit {
		alerts = append(alerts, RiskAlert{
			Type:           "HIGH_CONCENTRATION",
			Severity:       "MEDIUM",
			Message:        fmt.Sprintf("Position concentration %.1f%% exceeds limit", riskMetrics.ConcentrationRatio*100),
			Recommendation: "Diversify positions across multiple contracts",
		})
	}

	if riskMetrics.UnrealizedPnL < -riskMetrics.TotalExposure*0.1 {
		alerts = append(alerts, RiskAlert{
			Type:           "LARGE_UNREALIZED_LOSS",
			Severity:       "HIGH",
			Message:        "Unrealized losses exceed 10% of exposure",
			Recommendation: "Review positions and consider stop-loss orders",
		})
	}

	return alerts
}

// calculateRiskScore calculates overall risk score (0-100, higher is riskier)
func (rm *RiskManager) calculateRiskScore(marginStatus MarginStatus, varMetrics VaRMetrics, postTradeRisk map[string]interface{}) float64 {
	score := 0.0

	// Margin utilization component (0-40 points)
	score += math.Min(40, marginStatus.MarginUtilization*50)

	// VaR component (0-30 points)
	var varRatio float64
	if varMetrics.TotalExposure > 0 {
		varRatio = varMetrics.VaR / varMetrics.TotalExposure
	}
	score += math.Min(30, varRatio*1000)

	// Risk metrics component (0-30 points)
	if riskMetrics, ok := postTradeRisk["risk_metrics"].(RiskMetrics); ok {
		score += math.Min(15, riskMetrics.LeverageRatio*3)
		score += math.Min(15, riskMetrics.ConcentrationRatio*50)
	}

	return math.Min(100, score)
}

// generateRiskRecommendations generates risk management recommendations
func (rm *RiskManager) generateRiskRecommendations(riskScore float64, marginStatus MarginStatus, varMetrics VaRMetrics) []string {
	var recommendations []string

	if riskScore > 70 {
		recommendations = append(recommendations, "HIGH RISK: Immediate action required to reduce risk exposure")
	} else if riskScore > 50 {
		recommendations = append(recommendations, "MEDIUM RISK: Monitor positions closely and consider risk reduction")
	}

	if marginStatus.MarginUtilization > 0.8 {
		recommendations = append(recommendations, "Consider adding funds or reducing positions to improve margin situation")
	}

	if varMetrics.VaR > 10000 {
		recommendations = append(recommendations, "High VaR detected - consider portfolio hedging strategies")
	}

	return recommendations
}